src/rpc/services/SpectateService.ts

```diff
@@ -0,0 +1,418 @@
+import { ARpcService, MainRpcService } from "@/rpc";
+import { ChannelId, Timestamp } from "@/types/types";
+import { StateSnapshot } from "@/models";
+import Clock from "@/Clock";
+import ATransport from "@/transport/ATransport";
+import {
+    MilestoneProofStruct,
+    StateProofStruct
+} from "@typechain-types/contracts/V1/types/ProofTypes";
+import { DisputeStruct } from "@typechain-types/contracts/V1/types/DisputeTypes";
+import { Codec, Type } from "@/utils";
+import { isEqual } from "lodash";
+
+export interface SnapshotPayload {
+    disputeWindows: DisputeStruct[];
+    stateProof: StateProofStruct;
+    latestForkGenesisSnapshot: StateSnapshot;
+}
+
+class SpectateService extends ARpcService {
+    constructor(mainRpcService: MainRpcService) {
+        super(mainRpcService);
+    }
+
+    // Called locally to initiate spectate sync
+    public spectateSync(transport: ATransport, channelId: ChannelId) {
+        console.log("spectateSync !");
+        let time = Clock.getTimeInSeconds();
+        this.mainRpcService.rpcProxy
+            .onSpectateRequest(channelId, time)
+            .sendOne(transport);
+    }
+
+    public async onSpectateRequest(channelId: ChannelId, time: Timestamp) {
+        let localTime = Clock.getTimeInSeconds();
+        if (
+            Math.abs(time - localTime) >
+            this.mainRpcService.p2pManager.stateManager.timeConfig.agreementTime
+        ) {
+            console.log(
+                `onSpectateRequest - time difference too big - time:${time} localTime:${localTime} diff:${
+                    time - localTime
+                } aggreeTime:${
+                    this.mainRpcService.p2pManager.stateManager.timeConfig
+                        .agreementTime
+                }`
+            );
+            return;
+        }
+        console.log(`onSpectateRequest - localTime:${localTime} time:${time}`);
+
+        // Generate payload to prove the latest possible snapshot
+        // (but don't send it on-chain - send it to the spectator)
+        const snapshotPayload = await this.generateSnapshotPayload(channelId);
+
+        console.log(`onSpectateRequest - done`);
+        this.mainRpcService.rpcProxy
+            .onSpectateResponse(channelId, snapshotPayload, localTime)
+            .sendOne(this.mainRpcService.senderTransport!);
+    }
+
+    public async onSpectateResponse(
+        channelId: ChannelId,
+        snapshotPayload: SnapshotPayload,
+        responseTime: Timestamp
+    ) {
+        console.log(`onSpectateResponse - start`);
+        let localTime = Clock.getTimeInSeconds();
+        let rtt = localTime - responseTime;
+        if (
+            rtt >
+            this.mainRpcService.p2pManager.stateManager.timeConfig.agreementTime
+        ) {
+            console.log("onSpectateResponse - RTT too high, ignoring");
+            return;
```

Comment:
you want to abort this channel all together - disconnect from everyone if you detect at least 1 dishonest peer